Add CRT shaders

Clownacy opened the pull request:
I've ported a couple of shaders from RetroArch - particularly the crt-pi and crt-lottes-fast ones. Unlike other shaders such as crt-royale, these shaders are single-pass, allowing them to work within Dolphin's framework. These look pretty good when paired with native resolution and SSAA.

crt-lottes-fast:
![image](https://github.com/dolphin-emu/dolphin/assets/7667009/8108b7e1-d140-4d26-bc84-6b612b85ee8f)

unfiltered:
![image](https://github.com/dolphin-emu/dolphin/assets/7667009/5483d46a-4e15-4062-bb14-adb7c8d40d2b)

crt-pi:
![image](https://github.com/dolphin-emu/dolphin/assets/7667009/9cfd1152-908c-4d42-94e0-908a1fa83c1f)

Unfortunately, I only noticed #11888 after I had already ported crt-pi. Still, crt-pi is specifically optimised for lower-end hardware (Raspberry Pis, in particular), so perhaps these shaders still have some utility.

Various settings have been exposed through Dolphin's configuration API, crt-pi in particular boasts extra options that even the RetroArch version does not due to being hardcoded within the GLSL code itself.

Both crt-pi and crt-lottes-fast run with every backend that I have tested (OpenGL, Vulkan, Direct3D 11, and Direct3D 12).

Dogway commented:
Hi @Clownacy my shader is only for color correction + color management not scanlines per se so it can be run at increased resolutions which I prefer for GC and Wii games.

Still if there are people that prefer native+scanlines your ports are a good option. I recently found zfast_crt to be both very good while at the same time very fast. I used it to create a monolithic shader for the Chromecast (so figure how fast it is), this is my [version](https://github.com/Dogway/emulation-random/blob/master/RetroArch/Shaders/zfast_crt_geo.glsl) for RA with vignette, distortion, corner size, and phosphor*temperature. It still runs at around 40fps on Chromecast but will hardcode/optimize further.

Clownacy commented:
Thanks for the clarification!

nyanpasu64 commented:
Rather late to say this, but these look more like LCD shaders than CRT, because the phosphor grid is as wide as the scanlines are tall. In CRTs generally the phosphor spacing is narrower than the electron beam's width at full brightness (though the electron beam often gets narrower in dimly lit parts of the image).

golivax commented:
I personally love this! The crt-lottes-fast shader looks gorgeous on my 4k tv. Can't wait to see other high-quality shaders (e.g., zfast_crt) ported to Dolphin. Hopefully someday Dolphin will also support multi-pass shaders (e.g., royale, crt-guest-advanced, ntsc adaptive).🤞

golivax commented:
Hi all! Is there any chance we can get this merged? This PR is "just" porting two well-known crt shaders and both work. I understand Dolphin has its roadmap and that review time is limited, but this one shouldn't take long I believe. Thanks!

AdmiralCurtiss commented:
I'm never sure what to do with shader PRs since, honestly, I don't think they should be part of the Dolphin codebase at all, and if someone wants them they can just download them from some user-managed repository/forum thread/whatever of shaders and drop them into their Dolphin user folder. None of the maintainers (including myself) seem to care much about these, and those users that do can certainly do a much better job of curating the shaders than us.

walterg74 commented:
Hi @Clownacy , I am testing these out with the latest Dolphin, and neither works. I get this error:

![Dolphin_Shader_Error](https://github.com/dolphin-emu/dolphin/assets/100450406/a63aebbb-80dd-46d8-afe7-8816d3b224b4)

Clownacy commented:
@walterg74 Are you sure you downloaded the shaders correctly? The error message suggests that the first line of the shader's code is at least 124 characters long, but that should not be the case.

Here are links to the raw files. If you download these, they should be correct.
https://raw.githubusercontent.com/dolphin-emu/dolphin/b1537ca2326114087f7af7836f2a5ba54b764e6d/Data/Sys/Shaders/crt_lottes_fast.glsl
https://raw.githubusercontent.com/dolphin-emu/dolphin/b1537ca2326114087f7af7836f2a5ba54b764e6d/Data/Sys/Shaders/crt_pi.glsl

walterg74 commented:
@Clownacy I believe so, but I will download again from these links and check!

Kingedeji commented:
@Clownacy Thanks for the shaders and with all those settings it's perfect. Really good to have something like this in dolphin proper!

sebasasa commented:
I tried crt-pi on dolphin android, and i got an error
1:316: S0012: Global variable initializer must be a constant expression

I asume this probably stems (Like most problems of these kind) from android being limited by OpelGL ES? Or is there something else going on here? If it is indeed a GL version problem i could probably take the time to fix it if people are interested in giving that retro feel to portable gc gaming

Clownacy commented:
@sebasasa I have fixed crt-pi and tested it in both OpenGL ES and Vulkan on my Pixel 6. All should be working now.

Clownacy commented:
[My blog post](https://clownacy.wordpress.com/2023/06/30/porting-crt-shaders-from-retroarch-to-dolphin/), which goes over the process of porting these shaders, has consistently been the most viewed and commented-on of all my posts for the last two years, and is one of the top search results on Google for 'dolphin crt shader' (along with this very issue).

I think there is a sizeable demand for Dolphin to have a CRT shader, and merging this PR would make such a thing more accessible to users, saving them the need to navigate GitHub's arcane UI and Dolphin's internals in order to install it themselves.

sepalani commented:
@dolphin-emu-bot rebuild

sepalani commented:
@Clownacy
Could you please rebase your PR on latest master and squash your commits into a single one? The buildbot is complaining that your branch is too out-of-date.

I have no strong opinion regarding shaders. At the very least I can give it a try but I'm not confident about reviewing the actual code as it's outside my area of expertise.

Clownacy committed:
Add CRT shaders

I've ported a couple of shaders from RetroArch - particularly the
crt-pi and crt-lottes-fast ones. Unlike other shaders such as
crt-royale, these shaders are single-pass, allowing them to work
within Dolphin's framework. These look pretty good when paired with
native resolution and SSAA.

Clownacy commented:
Forgot how to set a branch's upstream correctly for a second there, but it should be correct now; all one commit, based on the latest master.

sepalani commented:
@dolphin-emu-bot rebuild

iwubcode commented:
As the main dev focused on graphics enhancements, I'll just leave some general ramblings on why I haven't pushed for these PRs to be merged.

---------

**(1)**  Admiral already touched on this way back when but I share the sentiment.

I want to promote users/communities working together and sharing their contributions.  So I love that this is being shared on a blog post and that people are using it that way.  I do think Dolphin needs to make it easier to share shaders/texture-packs/etc and that's a long term goal of mine to improve that process.

Because the process is a bit of a pain for users, I see why there is value to put the shader in Dolphin.  The negative is that once we put it in Dolphin, it no longer becomes the community's responsibility to keep it updated or to fix issues, it becomes Dolphin developers' responsibility.  If someone on the team was passionate about CRT, I think this would have been merged sooner, but no one on the team feels strongly about it.

**(2)** I have plans to update Dolphin's post processing system.

Dolphin's current system has a lot of short comings.  The most obvious is its lack of a multiple-pass system but there are a number of other issues as well.  I addressed many of these short comings in an old PR ( #10362 ) but even that has some issues due to how Dolphin operated at the time.

What I've been working toward the last few years is building an overall infrastructure for graphics enhancements.  This includes:  support for how shaders are loaded,  enhancing the rendering system to allow for other capabilities (like multi-pass compute shaders), determining how post processing fits in with the rest of the enhancements, etc.  This has been a huge task and it's not done yet.  There's many parts of this (post processing is just a tiny piece) and so that is why it has taken so long.  Users may not care about all these other features or how it fits into the big picture...but the goal is something far grander than a single shader.

So my main reason for not merging this has been that I'm trying to save myself some work in the future.  While it might be minor overall, I'd prefer to not go and update any old shaders to the new system.

**(3)**  This is very minor but I haven't thought of how to handle CRT shaders.

There are so many variants and it seems like the most popular ones (ex: CRT Royale) are not even possible until (2) is done.

If we provide multiple CRT shaders, how does a casual user know which one is worthwhile?  I suppose they can try them out but Dolphin lacks a lot of features that make it easy to see what is available in that space and what each does (how would I know crt-pi is for low end systems for instance?).

Lottes seems ok to my untrained eye but even on this thread there are some users that aren't happy with this choice.  Is there a universally accepted CRT shader?  If so, we'd be better off with that and let tech savvy users seek out other alternatives for specialized scenarios.

---------

All that said, I don't have any strong objections with adding this to Dolphin if someone else on the team gives it the thumbs up.

I will say to keep in mind if (2) happens (it's still hypothetical at this point, I've been promising it for years!), I probably won't bother to convert these shaders to the new system.  If someone wants them, they can re-add them...or add something more popular like CRT Royale.

iwubcode commented:
> I think what would help with a lot of this would be for Dolphin to be more opinionated. I agree that [an over-saturation of options is bad](https://ometer.com/preferences) for both the user and developer experience, so, if a CRT shader were to be accepted, then choosing only one (and not overloading the user with a barrage of hit-or-miss alternatives) would be best.

I think the opinion is that most of the team just doesn't care about shaders (feel free to correct me, team!).  Especially when users have an option to provide shaders amongst themselves.  The only advantage it provides is that it slightly improves accessibility but at the cost of maintenance burden, overhead of reviewing/approving more PRs, etc.

I will reiterate that I do care but there's only one of me.  And I'd rather wait for a high quality system with an equally exciting CRT shader as our first release.

> The lack of opinion is what has led to the current awkward situation, where Dolphin is preventing users from conveniently using useful and sought-after shaders whilst also bundling needless gimmick shaders like 'spooky' and 'acid'. This does not convey to users (and developers) that shaders are strictly third-party extensions which Dolphin will never officially provide, and it also creates the maintenance burden of needing to maintain and modernise all of those gimmick shaders in the future. Had Dolphin taken an official stance against providing shaders previously, both of these problems would have been avoided.

I don't like the gimmick shaders either.  You have to realize that these were added back in 2009.  The post processing system had [just come out](https://github.com/dolphin-emu/dolphin/commit/085fa5f526522674909f670ffa89c8b3f51bdf8d) and I can only assume people were somewhat excited to play with the system.  Another assumption would be that the devs didn't care as much back then or think of the ramifications of the bloat of adding a bunch of worthless shaders.

If/when the new system is created, I'd be all for removing the ones that the community doesn't like.  I think we have at most 3 or 4 shaders that are worthwhile...

> Instead of striving to provide (and develop!) the ultimate CRT shader, Dolphin could simply provide what its maintainers deem to be the best one currently available, after which they can simply forget about it aside from typical maintenance.

Yes, it's that maintenance that is the concern.  Even on your blog post, you mention Android issues in the comments which had to be addressed.

If you're ok with these shaders being broken (or missing) in the future, then I don't mind moving this forward.  I still think it'd be better to have the community share shaders but am not opposed to having "popular" shaders being added to Dolphin for accessibility reasons as long as there is no maintenance cost.

> In addition, I prefer CRT Royale to both of these shaders, so if/when support for multi-pass shaders is added, I would be greatly in favour of these two shaders being deprecated and removed; I only see them as stepping stones to a better solution in the future rather than permanent additions to Dolphin.

Alright, that's good enough for me.  This will break any user config but that's not going to change if they are using community driven shaders either.

iwubcode reviewed (approved):
Tested multiple backends.  It works!

CrazySeanDX commented:
I'm kinda with Jordan. If it's okay, how do I remove the curvature of the filter? I don't mind editing it myself, as I don't wanna pester you about changing it after 2 years.

Clownacy commented:
> I'm kinda with Jordan. If it's okay, how do I remove the curvature of the filter? I don't mind editing it myself, as I don't wanna pester you about changing it after 2 years.

It can be changed in Dolphin's shader configuration menu; right next to the drop-down menu for selecting the shader is a 'Configure' button. The curvature can be disabled by dragging the Corner Rounding and Curvature sliders to 0.

Clownacy commented:
Personally I prefer the shader with curvature, but I don't know how the majority feels about it. Googling the subject isn't helping me much. Curvature is enabled by default in [the upstream libretro shader](https://github.com/libretro/glsl-shaders/blob/a8136d8b8b5c6375296f833e7e81efa15ed76f11/crt/shaders/crt-lottes-fast.glsl#L101), so that may be some indication.

Clownacy commented:
Now that I think about it, the main point of porting these shaders was to make native resolution look good, and curvature doesn't really have anything to do with that. Curvature is disabled by default now, but can still be re-enabled in the settings.

sepalani commented:
@dolphin-emu-bot rebuild

jordan-woyak commented:
> Hey, so, dumb question. Can these be added to the Android version of Dolphin, or is that not doable?

They will be available in Android in the Enhancements menu with all the other shaders.

CrazySeanDX commented:
> > Hey, so, dumb question. Can these be added to the Android version of Dolphin, or is that not doable?
> 
> They will be available in Android in the Enhancements menu with all the other shaders.

In my experience, they're not. I don't know where to put them, if I'm even able to put them anywhere for it to show up.

pressRtowin commented:
> Now that I think about it, the main point of porting these shaders was to make native resolution look good, and curvature doesn't really have anything to do with that. Curvature is disabled by default now, but can still be re-enabled in the settings.

Just wanted to point out that as of now, the Android version of Dolphin does not appear to have a shader customization menu option, so the question about what most people are looking for in a CRT shader is something to still keep in mind, unless the Dolphin team is planning on adding this option to the Android UI in the near future.

Sawtaytoes commented:
What's the Status on these? The first thing I wanted is some sort of scanline/CRT shader because the games don't look right on my OLED.
